syntax = "proto3";

package judge;

option go_package = "/pb/judger";

message Config {
  int32 max_cpu_time = 1;
  int32 max_real_time = 2;
  int64 max_memory = 3;
  int64 max_stack = 4;
  int32 max_process_number = 5;
  int64 max_output_size = 6;
  int32 memory_limit_check_only = 7;
  string exe_path = 8;
  string input_path = 9;
  string output_path = 10;
  string error_path = 11;
  repeated string args = 12;
  repeated string env = 13;
  string log_path = 14;
  uint32 uid = 16;
  sint32 gid = 17;
  bool use_seccomp = 18;
}

enum ErrorCode {
  SUCCESS = 0;
  INVALID_CONFIG = -1;
  FORK_FAILED = -2;
  PTHREAD_FAILED = -3;
  WAIT_FAILED = -4;
  ROOT_REQUIRED = -5;
  LOAD_SECCOMP_FAILED = -6;
  SETRLIMIT_FAILED = -7;
  DUP2_FAILED = -8;
  SETUID_FAILED = -9;
  EXECVE_FAILED = -10;
  SPJ_ERROR = -11;
}

enum ResultCode {
  ACCEPTED = 0;
  WRONG_ANSWER = -1;
  CPU_TIME_LIMIT_EXCEEDED = 1;
  REAL_TIME_LIMIT_EXCEEDED = 2;
  MEMORY_LIMIT_EXCEEDED = 3;
  RUNTIME_ERROR = 4;
  SYSTEM_ERROR = 5;
};

message Result {
  int32 cpu_time = 1;
  int32 real_time = 2;
  int64 memory = 3;
  int32 signal = 4;
  int32 exit_code = 5;
  ErrorCode error = 6;
  ResultCode result = 7;
}